packages/core/src/sync/index.ts

```diff
@@ -1,318 +1,592 @@
+  for (const network of args.networks) {
+    if (network.disableCache) {
+      const startBlock = hexToNumber(
+        localSyncs.get(network)!.startBlock.number,
+      );
+
+      args.common.logger.warn({
+        service: "sync",
+        msg: `Deleting cache records for '${network.name}' from block ${startBlock}`,
+      });
+
+      await args.syncStore.pruneByChain({
+        fromBlock: startBlock,
+        chainId: network.chainId,
+      });
+    }
+  }
+
+  /** Convert `block` to a `Checkpoint`. */
+  const blockToCheckpoint = (
+    block: LightBlock | SyncBlock,
+    chainId: number,
+    rounding: "up" | "down",
+  ): Checkpoint => {
+    return {
+      ...(rounding === "up" ? maxCheckpoint : zeroCheckpoint),
+      blockTimestamp: hexToNumber(block.timestamp),
+      chainId: BigInt(chainId),
+      blockNumber: hexToBigInt(block.number),
+    };
   };
-  blockHash: Hex;
-  blockNumber: Hex;
-  result: {
-    address: Address;
-    code: Hex;
-    gasUsed: Hex;
-  } | null;
-  subtraces: number;
-  traceAddress: number[];
-  transactionHash: Hex;
-  transactionPosition: number;
-  type: "create";
-};
-export type SyncSuicideTrace = {
-  action: {
-    address: Address;
-    refundAddress: Address;
-    balance: Hex;
+
+  /**
+   * Returns the minimum checkpoint across all chains.
+   *
+   * Note: `localSync.latestBlock` is assumed to be defined if
+   * this function is called with `tag`: "latest".
+   */
+  const getChainsCheckpoint = <
+    tag extends "start" | "latest" | "finalized" | "end",
+  >(
+    tag: tag,
+  ): string | undefined => {
+    if (
+      tag === "end" &&
+      [...localSyncs.values()].some(
+        (localSync) => localSync.endBlock === undefined,
+      )
+    ) {
+      return undefined;
+    }
+
+    const checkpoints = [...localSyncs.entries()]
+      // Remove completed networks from checkpoint calculation for "latest"
+      .filter(
+        ([, localSync]) => tag !== "latest" || localSync.isComplete() === false,
+      )
+      .map(([network, localSync]) => {
+        const block = localSync[`${tag}Block`]!;
+
+        // The checkpoint returned by this function is meant to be used in
+        // a closed interval (includes endpoints), so "start" should be inclusive.
+        return blockToCheckpoint(
+          block,
+          network.chainId,
+          tag === "start" ? "down" : "up",
+        );
+      });
+
+    // Return `true` if all networks have been filtered out
+    if (checkpoints.length === 0) return undefined;
+    return encodeCheckpoint(checkpointMin(...checkpoints));
   };
-  blockHash: Hex;
-  blockNumber: Hex;
-  result: null;
-  subtraces: number;
-  traceAddress: number[];
-  transactionHash: Hex;
-  transactionPosition: number;
-  type: "suicide";
-};
-export type SyncRewardTrace = {
-  action: {
-    author: Address;
-    rewardType: "block" | "uncle";
-    value: Hex;
+
+  /** Updates `status` to record progress for each network. */
```

Comment:
Might be useful to have a longer comment here explaining the "fallback" logic, eg
```
1. If historical indexing is complete, use the most recent block present in the realtime sync with a checkpoint less than or equal to the overall indexed checkpoint.
2. If during historical indexing, use the most recent block present in the latest batch of events. If there are no events in the batch for a given network, the status block does not get updated for that network. It will look like that network is behind, but it actually just means there weren't any events.
3. ???
```

Reply:
Did you see the inline comments